add(doc): features list (#372)
chdsbd/kodiak: docs/docs/features.md

---
id: features
title: Features
sidebar_label: Features
---

This is a close to exhaustive list of Kodiak's features.

> If you have any questions about Kodiak, some good places to look are this page, the [configuration reference](config-reference.md), our [Kodiak recipes](recipes.md), and the [the GitHub repository](https://github.com/chdsbd/kodiak).

Reviewer: ```suggestion
If you have any questions about Kodiak, some good places to look are this page, the [configuration reference](config-reference.md), our [Kodiak recipes](recipes.md), and [the GitHub repository](https://github.com/chdsbd/kodiak).
```

## Merging Pull Requests

When the automerge label (configurable via [`merge.automerge_label`](https://kodiakhq.com/docs/config-reference#mergeautomerge_label)) is applied to a pull request, Kodiak will merge the pull request if all GitHub Branch Protection requirements are met (status checks, reviews, etc.).

You can disable the `merge.automerge_label` requirement by disabling `merge.require_automerge_label`. This means that as soon as your pull request meets the GitHub Branch Protection requirements, Kodiak will merge the pull request.

Some required GitHub status checks may remain pending indefinitely. This is common for status checks that require manual approval or GitHub apps like the [WIP GitHub App](https://github.com/wip/app), which sets a pending status check until the PR title is updated.
To prevent these status checks from blocking Kodiak's merge queue you must add them to `merge.dont_wait_on_status_checks`.

<!-- TODO: Link up all these config references -->

### Merge Methods

Like the GitHub UI, Kodiak supports merging via "squash", "rebase", and "merge" commits. This is configurable via [`merge.method`](https://kodiakhq.com/docs/config-reference#mergemethod).

> Kodiak also supports the "Require signed commits" branch protection setting for "squash" and "merge" commit methods. The "rebase" method is not compatible due to GitHub limitations.

Reviewer: ```suggestion
Kodiak also supports the "Require signed commits" branch protection setting for "squash" and "merge" commit methods. The "rebase" method is not compatible due to GitHub API limitations.
```

![Branch Protection – require signed commits](/img/branch-protection-require-signed-commits.png)

### Merge Message

The commit title and body of a comment can be customized with Kodiak via the `merge.message` configuration options.

- `merge.message.body` – use GitHub default body, the pull request body, or an empty string
- `merge.message.include_pr_number` – include the pull request number in the title like GitHub does

> - `merge.message.strip_html_comments` – remove HTML comments from the pull request body. This is good for remove pull request templates

Reviewer: ```suggestion
- `merge.message.strip_html_comments` – remove HTML comments from the pull request body. This is good for removing pull request templates
```

- `merge.message.include_pull_request_url` – include the pull request URL at the bottom of the commit message

### Preventing Merge

With GitHub Branch Protection, required status checks will prevent a pull request from merging until tests have passed. Draft pull requests will also prevent a pull request from merging.

Kodiak has some internal features to disable merging as well:

- `merge.blacklist_title_regex` – disable Kodiak merging based on a regex matching to the pull request title
- `merge.blacklist_labels` – disable Kodiak merging based on labels applied to the PR
- `merge.do_not_merge` – completely disable Kodiak from merging any PR. This is useful if you only want to use Kodiak for updating pull requests.

### Efficient Merging

With default settings, Kodiak will conserve continuous integration (CI) resources and only update a PR when necessary prior to merge. This can be altered with `update.always` at a cost of more CI load.

By default Kodiak's internal merge queue acts on a first-come, first-served basis. The first pull request that enters the queue will be merge before any other pull request, even if another pull request requires fewer branch updates.

> If `merge.prioritize_ready_to_merge` is enabled, if a PR is able to be merged without any branch updates, Kodiak will merge it, ignoring any pull requests in the queue.

Reviewer: ```suggestion
If `merge.prioritize_ready_to_merge` is enabled and a PR is able to be merged without any branch updates, then Kodiak will merge the PR ignoring any pull requests in the queue.
```
